File: desktop-viewer/platform-core/adb-core/src/test/kotlin/com/androidperformancestudio/platform/adb/AdbClientTest.kt
package com.androidperformancestudio.platform.adb

import com.androidperformancestudio.platform.toolchain.HostProcessBinaryResult
import com.androidperformancestudio.platform.toolchain.HostProcessLaunchRequest
import com.androidperformancestudio.platform.toolchain.HostProcessRequest
import com.androidperformancestudio.platform.toolchain.HostProcessRunner
import com.androidperformancestudio.platform.toolchain.HostProcessTextResult
import com.androidperformancestudio.platform.toolchain.RunningHostProcess
import kotlinx.coroutines.runBlocking
import java.nio.file.Path
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertIs
import kotlin.time.Duration

class AdbClientTest {
    @Test
    fun `discovers immutable device targets through the shared client`() =
        runBlocking {
            val runner =
                RecordingRunner(
                    textStdout =
                        "List of devices attached\n" +
                            "emulator-5554 device product:sdk model:Pixel_8 transport_id:1\n",
                )

            val target: DeviceTarget = DefaultAdbClient(Path.of("/sdk/adb"), runner).listDevices().single()

            assertIs<DeviceTarget>(target)
            assertEquals("emulator-5554", target.serial)
            assertEquals(AdbDeviceState.ONLINE, target.state)
        }

    @Test
    fun `preserves shell arguments and builds exec-out and forwarding arguments`() =
        runBlocking {
            val runner = RecordingRunner()
            val client = DefaultAdbClient(Path.of("/sdk/adb"), runner)

            client.shell("emulator-5554", listOf("getprop", "ro.build.type"))
            client.shell("emulator-5554", listOf("sh", "-c", "cat config | perfetto -c - --background-wait"))
            client.execOut("emulator-5554", listOf("screencap", "-p"))
            client.forward("emulator-5554", "tcp:39123", "localabstract:agentperf")
            client.removeForward("emulator-5554", "tcp:39123")

            assertEquals(
                listOf(
                    listOf("-s", "emulator-5554", "shell", "'getprop'", "'ro.build.type'"),
                    listOf(
                        "-s",
                        "emulator-5554",
                        "shell",
                        "'sh'",
                        "'-c'",
                        "'cat config | perfetto -c - --background-wait'",
                    ),
                    listOf("-s", "emulator-5554", "exec-out", "screencap", "-p"),
                    listOf("-s", "emulator-5554", "forward", "tcp:39123", "localabstract:agentperf"),
                    listOf("-s", "emulator-5554", "forward", "--remove", "tcp:39123"),
                ),
                runner.commands.map(HostProcessRequest::arguments),
            )
        }

    @Test
    fun `rejects unsafe serial and remote path before execution`() =
        runBlocking {
            val runner = RecordingRunner()
            val client = DefaultAdbClient(Path.of("/sdk/adb"), runner)

            assertFailsWith<AdbInputException> {
                client.shell("device; reboot", listOf("id"))
            }
            assertFailsWith<AdbInputException> {
                client.pull("device", "/data/local/tmp/\u0000trace", Path.of("trace"))
            }
            assertEquals(emptyList(), runner.commands)
        }

    @Test
    fun `propagates capture cancellation and output bounds through the typed boundary`() =
        runBlocking {
            val runner = RecordingRunner()
            val client = DefaultAdbClient(Path.of("/sdk/adb"), runner)

            client.shell(
                serial = "emulator-5554",
                arguments = listOf("id"),
                maxOutputBytesPerStream = 1024,
                isCancellationRequested = { true },
            )

            val request = runner.commands.single()
            assertEquals(1024, request.maxOutputBytesPerStream)
            assertEquals(true, request.isCancellationRequested())
        }

    private class RecordingRunner(
        private val textStdout: String = "",
    ) : HostProcessRunner {
        val commands = mutableListOf<HostProcessRequest>()

        override suspend fun executeText(request: HostProcessRequest): HostProcessTextResult {
            commands += request
            return HostProcessTextResult(-1, 0, textStdout, "", Duration.ZERO, false, false)
        }

        override suspend fun executeBinary(request: HostProcessRequest): HostProcessBinaryResult {
            commands += request
            return HostProcessBinaryResult(-1, 0, byteArrayOf(), byteArrayOf(), Duration.ZERO, false, false)
        }

        override fun launch(request: HostProcessLaunchRequest): RunningHostProcess =
            error("not used")
    }
}

private class AdbCoreMigrationTest {
    @Test
    fun `generic feature command executor keeps host execution inside adb core`() =
        runBlocking {
            val runner = AdbClientTestRecordingRunner(textStdout = "ok")
            val executor = DefaultAdbCommandExecutor(Path.of("/sdk/adb"), runner)

            val result = executor.executeText(listOf("-s", "serial-1", "shell", "run-as", "dev.example", "pwd"))

            assertEquals("ok", result.stdout)
            assertEquals(
                listOf("-s", "serial-1", "shell", "run-as", "dev.example", "pwd"),
                runner.commands.single().arguments,
            )
        }

    @Test
    fun `shared property client uses the precise adb core getprop command`() =
        runBlocking {
            val runner = AdbClientTestRecordingRunner(textStdout = "35\n")
            val properties = com.androidperformancestudio.adb.AndroidDevicePropertyClient(
                DefaultAdbClient(Path.of("/sdk/adb"), runner),
            )

            val result = properties.sdkInt("serial-1")

            assertEquals(35, (result as com.androidperformancestudio.model.StudioResult.Success).value)
            assertEquals(
                listOf("-s", "serial-1", "shell", "'getprop'", "'ro.build.version.sdk'"),
                runner.commands.single().arguments,
            )
        }

    @Test
    fun `shared command arguments validate and construct generic forwarding operations`() {
        assertEquals(
            listOf("-s", "serial-1", "forward", "tcp:0", "tcp:48123"),
            AdbCommandArguments.forward("serial-1", "tcp:0", "tcp:48123"),
        )
        assertEquals(
            listOf("-s", "serial-1", "forward", "--remove", "tcp:48123"),
            AdbCommandArguments.removeForward("serial-1", "tcp:48123"),
        )
        assertFailsWith<AdbInputException> {
            AdbCommandArguments.pidOf("serial-1", "dev.example; rm -rf /")
        }
    }
}

private class AdbClientTestRecordingRunner(
    private val textStdout: String,
) : HostProcessRunner {
    val commands = mutableListOf<HostProcessRequest>()

    override suspend fun executeText(request: HostProcessRequest): HostProcessTextResult {
        commands += request
        return HostProcessTextResult(-1, 0, textStdout, "", Duration.ZERO, false, false)
    }

    override suspend fun executeBinary(request: HostProcessRequest): HostProcessBinaryResult = error("not used")

    override fun launch(request: HostProcessLaunchRequest): RunningHostProcess = error("not used")
}

File: electron-viewer/packages/platform-adb/src/adb-client.test.ts
import { describe, expect, it } from 'vitest';
import type { HostProcessBinaryResult, HostProcessRequest } from '@aps/platform-host';
import { AdbClient, type AdbBinaryExecutor } from './adb-client.js';
import { AdbCommandFailedError, AdbInputError } from './adb-errors.js';

const DEVICES_OUTPUT = [
  'List of devices attached',
  'emulator-5554 device product:sdk model:Pixel device:emu transport_id:7',
].join('\n');

function fakeExecutor(stdout: string) {
  const calls: HostProcessRequest[] = [];
  const execute: AdbBinaryExecutor = async (request) => {
    calls.push(request);
    const result: HostProcessBinaryResult = {
      pid: 42,
      exitCode: 0,
      stdout: Buffer.from(stdout, 'utf8'),
      stderr: Buffer.alloc(0),
      durationMs: 1,
      stdoutTruncated: false,
      stderrTruncated: false,
    };
    return result;
  };
  return { calls, execute };
}

describe('AdbClient', () => {
  it('lists and parses devices', async () => {
    const fake = fakeExecutor(DEVICES_OUTPUT);
    const client = new AdbClient({ executable: '/sdk/adb', execute: fake.execute });
    const devices = await client.listDevices();
    expect(devices[0]).toMatchObject({ serial: 'emulator-5554', state: 'ONLINE', transportId: 7 });
    expect(fake.calls[0]?.args).toEqual(['devices', '-l']);
  });

  it('quotes remote shell arguments but not exec-out arguments', async () => {
    const shellFake = fakeExecutor('ok');
    const shellClient = new AdbClient({ executable: '/sdk/adb', execute: shellFake.execute });
    await shellClient.shell('SER', ['pm', 'path', "com.x'y"]);
    expect(shellFake.calls[0]?.args).toEqual([
      '-s',
      'SER',
      'shell',
      "'pm'",
      "'path'",
      "'com.x'\"'\"'y'",
    ]);

    const execFake = fakeExecutor('raw');
    const execClient = new AdbClient({ executable: '/sdk/adb', execute: execFake.execute });
    await execClient.execOut('SER', ['screencap', '-p']);
    expect(execFake.calls[0]?.args).toEqual(['-s', 'SER', 'exec-out', 'screencap', '-p']);
  });

  it('builds push, pull, forward, remove-forward, and bugreport argument vectors', async () => {
    const fake = fakeExecutor('ok');
    const client = new AdbClient({
      executable: '/sdk/adb',
      execute: fake.execute,
      isRegularFile: (path) => path === '/local/file',
    });
    await client.push('SER', '/local/file', '/data/local/tmp/f');
    await client.pull('SER', '/data/local/tmp/f', '/local/out');
    await client.forward('SER', 'tcp:0', 'localabstract:agent');
    await client.removeForward('SER', 'tcp:1234');
    await client.bugreport('SER', '/local/bugreport.zip');
    expect(fake.calls.map((call) => call.args)).toEqual([
      ['-s', 'SER', 'push', '/local/file', '/data/local/tmp/f'],
      ['-s', 'SER', 'pull', '/data/local/tmp/f', '/local/out'],
      ['-s', 'SER', 'forward', 'tcp:0', 'localabstract:agent'],
      ['-s', 'SER', 'forward', '--remove', 'tcp:1234'],
      ['-s', 'SER', 'bugreport', '/local/bugreport.zip'],
    ]);
  });

  it('rejects a push whose local source is not a regular file', async () => {
    const fake = fakeExecutor('ok');
    const client = new AdbClient({ executable: '/sdk/adb', execute: fake.execute, isRegularFile: () => false });
    await expect(client.push('SER', '/missing', '/remote')).rejects.toBeInstanceOf(AdbInputError);
  });

  it('fails on a non-zero exit code', async () => {
    const execute: AdbBinaryExecutor = async () => ({
      pid: 1,
      exitCode: 1,
      stdout: Buffer.alloc(0),
      stderr: Buffer.from('device offline', 'utf8'),
      durationMs: 1,
      stdoutTruncated: false,
      stderrTruncated: false,
    });
    const client = new AdbClient({ executable: '/sdk/adb', execute });
    await expect(client.listDevices()).rejects.toBeInstanceOf(AdbCommandFailedError);
  });
});
